src/org/javarosa/core/model/actions/SetValueAction.java

```diff
@@ -28,34 +28,45 @@
  */
 public class SetValueAction extends Action {
     private TreeReference target;
+    private TreeReference trigger;
     private XPathExpression value;
     private String explicitValue;
 
     public SetValueAction() {
 
     }
 
-    public SetValueAction(TreeReference target, XPathExpression value) {
+    public SetValueAction(TreeReference target,  TreeReference trigger, XPathExpression value) {
         super("setvalue");
         this.target = target;
         this.value = value;
+        this.trigger = trigger;
     }
 
-    public SetValueAction(TreeReference target, String explicitValue) {
+    public SetValueAction(TreeReference target, TreeReference trigger, String explicitValue) {
         super("setvalue");
         this.target = target;
         this.explicitValue = explicitValue;
+        this.trigger = trigger;
     }
 
     public void processAction(FormDef model, TreeReference contextRef) {
 
+        // If the trigger was specified then check if the context ref is not null and actually equals the trigger.
+        if (trigger != null && (contextRef == null || !trigger.equals(contextRef.genericize()))) {
+            return;
+        }
+
         //Qualify the reference if necessary
         TreeReference qualifiedReference = contextRef == null ? target : target.contextualize(contextRef);
 
         //For now we only process setValue actions which are within the
         //context if a context is provided. This happens for repeats where
         //insert events should only trigger on the right nodes
-        if(contextRef != null){
+        //If the trigger is specified then it means it's Action.EVENT_XFORMS_VALUE_CHANGED and thus
+        //it's not important if the target is withing the context. Context points to the node which value has changed and the target
+        //to be updated may be at any position within the form (instance).
+        if(contextRef != null && trigger == null) {
```

Comment:
Space after `if`
Do we have a space after `//` usually?

Reply:
I'll fix the space.

I don't think we usually have a space after `//` but I just followed the style of the file. 
But I prefer the space after to be clear :)